elements/research/ddiotune.cc

```diff
@@ -0,0 +1,385 @@
+/*
+ * DDIOTune.{cc,hh} --
+ * Alireza Farshin
+ *
+ * Copyright (c) 2020 KTH Royal Institute of Technology
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, subject to the conditions
+ * listed in the Click LICENSE file. These conditions include: you must
+ * preserve this copyright notice, and you cannot mention the copyright
+ * holders in advertising related to the Software without their permission.
+ * The Software is provided WITHOUT ANY WARRANTY, EXPRESS OR IMPLIED. This
+ * notice is a summary of the Click LICENSE file; the license in that file is
+ * legally binding.
+ */
+
+#include <click/config.h>
+#include <click/args.hh>
+#include <click/error.hh>
+#include <cpuid.h>
+#include <unistd.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <cmath>
+#include "ddiotune.hh"
+#include <rte_lcore.h>
+
+CLICK_DECLS
+
+/* Read from an MSR register */
+int32_t
+DDIOTune::rdmsr(int msr, uint64_t *val, int cpu)
+{
+	int fd;
+	int ret;
+	char msr_file[50];
+	sprintf(msr_file,"/dev/cpu/%d/msr",cpu);
+
+	fd = open(msr_file, O_RDONLY);
```

Comment:
Instead of issuing 2 syscalls (open and pread), you could simply use the native Click method: file_string()
This method returns a String, but you could easily convert the result to int.

Reply:
That would lead to multiple conversion and go through unnecessary null-termination. And the file content would appear as a weird quad word as string. If anything then add a function like file_read_uint64. But it's so specific that I think it's fine.

Reply:
If @tbarbette is okay with it, I will skip this :-p 
It will be called during the initialization phase, so 2 syscall is not a big issue.